crates/pipeline-manager/src/compiler/main.rs

```diff
@@ -99,6 +101,261 @@ async fn get_binary(
+                        acc.extend_from_slice(&chunk);
+                        Ok(acc)
+                    })
+                    .await
+                    .map_err(|e| {
+                        ManagerError::from(ApiError::InvalidMultipartData {
+                            error: format!("Failed to read filename: {}", e),
+                        })
+                    })?;
+                _filename = Some(String::from_utf8(bytes).map_err(|e| {
+                    ManagerError::from(ApiError::InvalidMultipartData {
+                        error: format!("Invalid UTF-8 in filename: {}", e),
+                    })
+                })?);
+            }
+            _ => {
+                warn!("Ignoring unknown multipart field: {}", field_name);
+            }
+        }
+    }
+
+    // Validate required fields
+    let pipeline_id = pipeline_id.ok_or_else(|| {
+        ManagerError::from(ApiError::MissingMultipartField {
+            field: "pipeline_id".to_string(),
+        })
+    })?;
+    let program_version = program_version.ok_or_else(|| {
+        ManagerError::from(ApiError::MissingMultipartField {
+            field: "program_version".to_string(),
+        })
+    })?;
+    let source_checksum = source_checksum.ok_or_else(|| {
+        ManagerError::from(ApiError::MissingMultipartField {
+            field: "source_checksum".to_string(),
+        })
+    })?;
+    let expected_integrity_checksum = integrity_checksum.ok_or_else(|| {
+        ManagerError::from(ApiError::MissingMultipartField {
+            field: "integrity_checksum".to_string(),
+        })
+    })?;
+
+    if binary_data.is_empty() {
+        return Err(ManagerError::from(ApiError::MissingMultipartField {
+            field: "binary".to_string(),
+        }));
+    }
+
+    // Validate parameters
+    let pipeline_id_uuid =
+        Uuid::from_str(&pipeline_id).map_err(|e| ApiError::InvalidUuidParam {
+            value: pipeline_id.clone(),
+            error: e.to_string(),
+        })?;
+    let pipeline_id = PipelineId(pipeline_id_uuid);
+
+    let program_version =
+        Version(
+            i64::from_str(&program_version).map_err(|e| ApiError::InvalidVersionParam {
+                value: program_version.clone(),
+                error: e.to_string(),
+            })?,
+        );
+
+    validate_is_sha256_checksum(&source_checksum).map_err(|e| {
+        ManagerError::from(ApiError::InvalidChecksumParam {
+            value: source_checksum.to_string(),
+            error: e,
+        })
+    })?;
+
+    validate_is_sha256_checksum(&expected_integrity_checksum).map_err(|e| {
+        ManagerError::from(ApiError::InvalidChecksumParam {
+            value: expected_integrity_checksum.to_string(),
+            error: e,
+        })
+    })?;
+
+    // Calculate actual integrity checksum of the received binary
+    let actual_integrity_checksum = hex::encode(sha256(&binary_data));
```

Comment:
same as comment above ideally we can not have everything in DRAM of possible, you can calcultae the checksum based on chunks, there is a function in utils